src/execution/join_hashtable.cpp

```diff
@@ -243,70 +445,267 @@ idx_t JoinHashTable::PrepareKeys(DataChunk &keys, vector<TupleDataVectorFormat>
+	}
+}
+
+template <bool PARALLEL>
+static void InsertHashesLoop(atomic<ht_entry_t> entries[], Vector row_locations, Vector &hashes_v, const idx_t &count,
+                             JoinHashTable::InsertState &state, unique_ptr<TupleDataCollection> &data_collection,
+                             JoinHashTable *ht) {
+
+	D_ASSERT(hashes_v.GetType().id() == LogicalType::HASH);
+
+	ApplyBitmaskAndGetSaltBuild(hashes_v, count, ht->bitmask);
+
+	// the offset for each row to insert
+	idx_t *ht_offsets_and_salts = FlatVector::GetData<idx_t>(hashes_v);
+	// the row locations of the rows that are already in the hash table
+	data_ptr_t *rhs_row_locations = FlatVector::GetData<data_ptr_t>(state.rhs_row_locations);
+	// the row locations of the rows that are to be inserted
+	data_ptr_t *lhs_row_locations = FlatVector::GetData<data_ptr_t>(row_locations);
+
+	const SelectionVector *remaining_sel = FlatVector::IncrementalSelectionVector();
+	idx_t remaining_count = count;
+
+	// use the ht bitmask to make the modulo operation faster but keep the salt bits intact
+	idx_t capacity_mask = ht->bitmask | ht_entry_t::SALT_MASK;
+
+	while (remaining_count > 0) {
+
+		idx_t salt_match_count = 0;
+
+		// iterate over each entry to find out whether it belongs to an existing list or will start
+		// a new list
+		for (idx_t i = 0; i < remaining_count; i++) {
+
+			const idx_t row_index = remaining_sel->get_index(i);
+
+			idx_t &ht_offset_and_salt = ht_offsets_and_salts[row_index];
+			const hash_t salt = ht_entry_t::ExtractSalt(ht_offset_and_salt);
+
+			idx_t ht_offset;
+			ht_entry_t entry;
+			bool occupied, salt_match;
+
+			// increment the ht_offset_and_salt of the entry as long as next entry is occupied and salt does not match
+			while (true) {
+				ht_offset = ht_offset_and_salt & ht_entry_t::POINTER_MASK;
+				atomic<ht_entry_t> &atomic_entry = entries[ht_offset];
+				entry = atomic_entry.load(std::memory_order_relaxed);
+				occupied = entry.IsOccupied();
+				salt_match = entry.GetSalt() == salt;
+
+				// condition for incrementing the ht_offset: occupied and row_salt does not match -> move to next entry
+				if (!occupied || salt_match) {
+					break;
+				}
+
+				IncrementAndWrap(ht_offset_and_salt, capacity_mask);
+			}
+
+			if (!occupied) { // insert into free
+				atomic<ht_entry_t> &atomic_entry = entries[ht_offset];
+				data_ptr_t row_ptr_to_insert = lhs_row_locations[row_index];
+				data_ptr_t potential_collided_ptr =
+				    InsertRowToEntry<PARALLEL, true>(atomic_entry, row_ptr_to_insert, salt, ht->pointer_offset);
+
+				if (PARALLEL) {
+					// if the insertion was not successful, the entry was occupied in the meantime, so we have to
+					// compare the keys and insert the row to the next entry
+					if (potential_collided_ptr) {
+						// if the entry was occupied, we need to compare the keys and insert the row to the next entry
+						// we need to compare the keys and insert the row to the next entry
```

Comment:
Would it be better to check salt first and compare key only when salt is equal? By the way this line is duplicated.

Reply:
Hi @xuke-hat, thank you for the review! Good idea, I'll try to add the rows that fall in the case of line 628 to the remaining selection to process that in the next run of the outer loop, that way they will first compare the salt and then potentially the keys. 
Also, I'll address the code duplicate.